File: graph_prims.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Weighted edge structure
struct Edge {
    int target;
    int weight;
};

// Function to add a weighted undirected edge
void addEdge(vector<vector<Edge>>& adj, int u, int v, int w) {
    adj[u].push_back({v, w});
    adj[v].push_back({u, w});
}

// Core Prim's Algorithm Function
void primsMST(const vector<vector<Edge>>& adj, int V) {
    // Priority Queue to store pair <weight, node>
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    // Track visited nodes which are included in MST
    vector<bool> inMST(V, false);
    
    int mst_weight = 0; // Total cost of MST

    // Node 0 se shuru karte hain. Shuruati cost 0 hoga.
    pq.push({0, 0});

    cout << "--- Edges Included in the Minimum Spanning Tree ---" << endl;

    while (!pq.empty()) {
        int w = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        // Agar node pehle se MST ka hissa hai, toh skip karein
        if (inMST[u]) continue;

        // Node ko MST mein shamil karein aur cost add karein
        inMST[u] = true;
        mst_weight += w;

        if (w != 0) {
            cout << "Included Edge with weight: " << w << " (Reached Node " << u << ")" << endl;
        }

        // Neighbors ko check karein
        for (const auto& edge : adj[u]) {
            int v = edge.target;
            int weight = edge.weight;

            // Agar neighbor abhi tak MST mein nahi hai, toh queue mein daalein
            if (!inMST[v]) {
                pq.push({weight, v});
            }
        }
    }

    cout << endl << "Total Weight of the Minimum Spanning Tree: " << mst_weight << endl;
    cout << "---------------------------------------------------" << endl;
}

int main() {
    cout << "--- DSA Masterclass: Prim's Minimum Spanning Tree ---" << endl << endl;

    int V = 4; // 4 Nodes (0 se 3)
    vector<vector<Edge>> adj(V);

    // Graph setup (Connecting nodes with weights)
    addEdge(adj, 0, 1, 10);
    addEdge(adj, 0, 2, 6);
    addEdge(adj, 0, 3, 5);
    addEdge(adj, 1, 3, 15);
    addEdge(adj, 2, 3, 4);

    // Call Prim's Algorithm
    primsMST(adj, V);

    return 0;
}

File: graph_prims.py
import heapq

class GraphPrims:
    def __init__(self, vertices):
        self.V = vertices
        # Adjacency list: हर नोड के लिए खाली लिस्ट बनाएंगे
        self.adj = {i: [] for i in range(vertices)}

    def add_edge(self, u, v, weight):
        # Un-directed ग्राफ है, इसलिए दोनों तरफ जोड़ेंगे
        self.adj[u].append((v, weight))
        self.adj[v].append((u, weight))

    def prims_mst(self):
        # Min-Heap: (weight, current_node, parent_node) स्टोर करेगा
        # हम 0 नोड से शुरू कर रहे हैं, जिसका शुरुआती वेट 0 है और पैरेंट None है
        min_heap = [(0, 0, None)]
        
        # यह ट्रैक करने के लिए कि कौन सा नोड MST में शामिल हो चुका है
        visited = [False] * self.V
        
        mst_edges = []
        total_mst_weight = 0

        print("--- Edges included in Minimum Spanning Tree (Prim's) ---")

        while min_heap:
            weight, u, parent = heapq.heappop(min_heap)

            # अगर नोड पहले से विज़िट हो चुका है, तो छोड़ दें
            if visited[u]:
                continue

            # नोड को MST में शामिल करें
            visited[u] = True
            total_mst_weight += weight

            # अगर यह सोर्स नोड नहीं है, तो इसकी एज को रिजल्ट में दिखाएं
            if parent is not None:
                print(f"Edge ({parent} - {u}) with weight: {weight}")

            # अब इस नोड के सभी पड़ोसियों को हीप में डालें
            for neighbor, edge_weight in self.adj[u]:
                if not visited[neighbor]:
                    heapq.heappush(min_heap, (edge_weight, neighbor, u))

        print(f"Total Weight of MST: {total_mst_weight}")

# --- Main Driver Code ---
if __name__ == "__main__":
    # 4 नोड्स वाला ग्राफ (0, 1, 2, 3)
    g = GraphPrims(4)

    g.add_edge(0, 1, 10)
    g.add_edge(0, 2, 6)
    g.add_edge(0, 3, 5)
    g.add_edge(1, 3, 15)
    g.add_edge(2, 3, 4)

    # Prim's MST रन करें
    g.prims_mst()